src/transport/tests/TestTCP.cpp

```diff
@@ -0,0 +1,216 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements unit tests for the TcpTransport implementation.
+ */
+
+#include "TestTransportLayer.h"
+
+#include "NetworkTestHelpers.h"
+
+#include <core/CHIPCore.h>
+#include <support/CodeUtils.h>
+#include <transport/TCP.h>
+
+#include <nlbyteorder.h>
+#include <nlunit-test.h>
+
+#include <errno.h>
+
+using namespace chip;
+
+static int Initialize(void * aContext);
+static int Finalize(void * aContext);
+
+namespace {
+
+constexpr size_t kMaxTcpActiveConnectionCount = 4;
+constexpr size_t kMaxTcpPendingPackets        = 4;
+
+using TCPImpl = Transport::TCP<kMaxTcpActiveConnectionCount, kMaxTcpPendingPackets>;
+
+constexpr NodeId kSourceNodeId      = 123654;
+constexpr NodeId kDestinationNodeId = 111222333;
+constexpr uint32_t kMessageId       = 18;
+
+using TestContext = chip::Test::IOContext;
+TestContext sContext;
+
+const char PAYLOAD[]        = "Hello!";
+int ReceiveHandlerCallCount = 0;
+
+void MessageReceiveHandler(MessageHeader & header, const Transport::PeerAddress & source, System::PacketBuffer * msgBuf,
```

Comment:
We really need to document the ownership API contracts for this stuff.  I would _assume_ that this should take ownership of the incoming buffer, freeing it before returning, like `SecureSessionMgrBase::HandleDataReceived`does.  But the API on `Transport::Base` does not actually document the expected behavior of the function that gets passed to `SetMessageReceiveHandler`, so maybe it's the secure session manager that's doing this wrong?

Reply:
At least for TCP, buffers may contain several messages so it is not correct for handlers to free the data. I will review the usages to see if this is maintained within UDP and such.